### ISSUE TITLE
IAM diffs include duplicates

### ISSUE THREAD

**Issue by marcb1** [terraform-providers/terraform-provider-google#4288]
Something's not right with the way the google provider sorts IAM diffs by role.

The following git diff change was made:
```diff
    "roles/editor" = [
      "user:user2@email.com",
-     "user:user1@email.com",
      "serviceAccount:transactions@appspot.gserviceaccount.com",
      "serviceAccount:xxxxxx@cloudservices.gserviceaccount.com",
      "serviceAccount:xxxxxx-compute@developer.gserviceaccount.com",
      "serviceAccount:service-xxxxxx@containerregistry.iam.gserviceaccount.com",
    ]
    "roles/iam.securityAdmin" = [
      "user:user1@email.com",
    ]
-   "roles/iam.serviceAccountTokenCreator" = [
-    "serviceAccount:service-xxxxxx@gcp-sa-pubsub.iam.gserviceaccount.com",
-  ]
```

The following IAM diff was produced, `roles/pubsub.subscriber` and `roles/owner` are listed twice in the diff which created some confusion. Though applying the terraform change resulted in no issues, so I'm assuming that the sorting of the diff output is broken.

```
  # module.transactions_gcp_project.google_project_iam_policy.iam_policy will be updated in-place
  ~ resource "google_project_iam_policy" "project_iam_policy" {
        etag        = "BwWPD4EM9Ks="
        id          =  "transactions"
      ~ policy_data = jsonencode(
          ~ {
              ~ bindings = [
                    {
                        members = [
                            "serviceAccount:service-xxxxxx@gcp-sa-automl.iam.gserviceaccount.com",
                        ]
                        role    = "roles/automl.serviceAgent"
                    },
                    {
                        members = [
                            "serviceAccount:service-xxxxxx@gcf-admin-robot.iam.gserviceaccount.com",
                        ]
                        role    = "roles/cloudfunctions.serviceAgent"
                    },
                    {
                        members = [
                            "serviceAccount:service-xxxxxx@compute-system.iam.gserviceaccount.com",
                        ]
                        role    = "roles/compute.serviceAgent"
                    },
                    {
                        members = [
                            "serviceAccount:service-xxxxxx@container-engine-robot.iam.gserviceaccount.com",
                        ]
                        role    = "roles/container.serviceAgent"
                    },
                    {
                        members = [
                            "serviceAccount:service-xxxxxx@dataflow-service-producer-prod.iam.gserviceaccount.com",
                        ]
                        role    = "roles/dataflow.serviceAgent"
                    },
                  ~ {
                      ~ members = [
                            "serviceAccount:xxxxxx-compute@developer.gserviceaccount.com",
                            "serviceAccount:xxxxxx@cloudservices.gserviceaccount.com",
                            "serviceAccount:service-xxxxxx@containerregistry.iam.gserviceaccount.com",
                            "serviceAccount:tf@appspot.gserviceaccount.com",
                          - "user:user1@email.com",
                            "user:user2@email.com",
                        ]
                        role    = "roles/editor"
                    },
                  ~ {
                      ~ members = [
                          - "user:user1@email.com",
                          + "serviceAccount:service-xxxxxx@gcp-sa-pubsub.iam.gserviceaccount.com",
                        ]
                      ~ role    = "roles/iam.securityAdmin" -> "roles/iam.serviceAccountTokenCreator"
                    },
                  + {
                      + members = [
                          + "serviceAccount:tf@transactions.iam.gserviceaccount.com",
                          + "user:user3@email.com",
                          + "user:user4@email.com",
                        ]
                      + role    = "roles/owner"
                    },
                  + {
                      + members = [
                          + "serviceAccount:android-ipn-pull@transactions.iam.gserviceaccount.com",
                          + "serviceAccount:iap-ipn-push-27763922960@transactions.iam.gserviceaccount.com",
                        ]
                      + role    = "roles/pubsub.subscriber"
                    },
                    {
                        members = [
                            "serviceAccount:service-27763922960@gcp-sa-pubsub.iam.gserviceaccount.com",
                        ]
                        role    = "roles/iam.serviceAccountTokenCreator"
                    },
                    {
                        members = [
                            "serviceAccount:tf@transactions.iam.gserviceaccount.com",
                            "user:user3@email.com",
                            "user:user4@email.com",
                        ]
                        role    = "roles/owner"
                    },
                    {
                        members = [
                            "serviceAccount:android-ipn-pull@transactions.iam.gserviceaccount.com",
                            "serviceAccount:iap-ipn-push-xxxxxx@transactions.iam.gserviceaccount.com",
                        ]
                        role    = "roles/pubsub.subscriber"
                    },
                ]
            }
        )
        project     = "transactions"
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by slevenick** [comment]
Hey @marcb1 

Can you share the version of both terraform and the google terraform provider that you are using?

**Comment by marcb1** [comment]
we are using the latest terraform versions:
```
$terraform version
Terraform v0.12.6
+ provider.google v2.10.0
```

**Comment by slevenick** [comment]
Given that the applied IAM policy was correct, this is a relatively low priority.

I believe the batching logic that was added in the most recent provider version 2.13.0 may have improved the diff feature for IAM. `project: IAM and service requests are now batched (#4207)
` 

Could you try updating to 2.13.0 and see if the diffs are any more useful?

**Comment by marcb1** [comment]
We are still seeing this and it's making our IAM diffs unreadable, we are on the following version:
```
Terraform v0.12.6
+ provider.google v2.17.0
+ provider.google-beta v2.17.0
```

I think the problem is the diff sorts the roles and ends up keeping duplicates if they happen to fall in between the diff'd items. For example in the original ticket description `roles/owner` is listed twice.

Any ideas if this can be fixed, I'm happy to fix it myself, looking at the original here https://github.com/terraform-providers/terraform-provider-google/pull/3855
I'm not sure where the two IAM policies are diff'd in the code

**Comment by slevenick** [comment]
I'm not sure how this is happening, are you able to provide a config that reproduces the issue?

[This](https://github.com/terraform-providers/terraform-provider-google/blob/b35d35b59f1582c3c64ca40ea64e61746cb90a7d/google/iam.go#L339) is where the comparison should be happening, but [where we create the map out of the list of bindings](https://github.com/terraform-providers/terraform-provider-google/blob/b35d35b59f1582c3c64ca40ea64e61746cb90a7d/google/iam.go#L192) to diff *should* deduplicate on the role, so I'm not sure why this is happening.

**Comment by marcb1** [comment]
ok, so I tracked this in the code a bit more and looks like the diffs are done by the terraform core library. The issue is a side effect of storing the diff output as a JSON schema. Found a good discussion of the issue here
https://github.com/hashicorp/terraform/issues/21473

Perhaps one way is to use a terraform schema for representing the IAM policy instead of a JSON string, if this is not feasible, I see a PR that could potentially fix the issue here, https://github.com/hashicorp/terraform/pull/22695/files I'll try again to see if it fixes the issue

**Comment by danawillow** [comment]
Looks like the upstream fix was merged as part of TF v0.12.14. @marcb1, can you give that a try?

**Comment by marcb1** [comment]
yes much better now! closing this out
